packages/server/src/handlers/BaseHandler.ts

```diff
@@ -101,4 +110,58 @@ export class BaseHandler extends EventEmitter {
 
     return {host: host as string, proto}
   }
+
+  protected getLocker(req: http.IncomingMessage) {
+    if (typeof this.options.locker === 'function') {
+      return this.options.locker(req)
+    }
+    return this.options.locker
+  }
+
+  protected async acquireLock(
+    req: http.IncomingMessage,
+    id: string,
+    context: CancellationContext
+  ) {
+    const locker = this.getLocker(req)
+    await locker?.lock(id, () => {
+      context.cancel()
+    })
+
+    return () => locker?.unlock(id)
+  }
+
+  protected writeToStore(
+    req: http.IncomingMessage,
+    id: string,
+    offset: number,
+    context: CancellationContext
+  ) {
+    return new Promise<number>(async (resolve, reject) => {
+      if (context.signal.aborted) {
+        reject(ERRORS.ABORTED)
+        return
+      }
+
+      const proxy = new stream.PassThrough()
+      stream.addAbortSignal(context.signal, proxy)
+
+      proxy.on('error', (err) => {
+        req.unpipe(proxy)
+        if (err.name === 'AbortError') {
+          reject(ERRORS.ABORTED)
+        } else {
+          reject(err)
+        }
+      })
+
+      req.on('error', (err) => {
+        if (!proxy.closed) {
+          proxy.destroy(err)
+        }
+      })
+
+      this.store.write(req.pipe(proxy), id, offset).then(resolve).catch(reject)
```

Comment:
Could `stream.pipeline` be used here to simplify the error handling?

Reply:
unfortunately not, since if we use stream.pipeline it would automatically call `req.destroy()` and we don't want that, we first want to return a response, then call req.destroy()

Reply:
that's why I'm using a `PassThrough` proxy and add to that stream the abortSignal. This is thought to prevent the pipeline calling `req.destroy()` which will absurdly close the connection

Reply:
Alright, thank you.